File: netlify/functions/lib/scrapers/torrentdownload.js
/**
 * TorrentDownload scraper - HTML scraping.
 */
const client = require('../client');
const { parseSeeders, parseSize, cleanName, makeMagnet } = require('../utils');

const NAME = 'TorrentDownload';
const BASE = 'https://www.torrentdownload.info';

async function search(query) {
  const url = `${BASE}/search?q=${encodeURIComponent(query)}`;
  const $ = await client.getSoup(url);
  if (!$) return [];

  const results = [];
  const table = $('table.table2');
  if (!table.length) return [];

  table.find('tr').slice(1).each((_, row) => {
    const cols = $(row).find('td');
    if (cols.length < 4) return;

    const nameTag = $(cols[0]).find('a').first();
    const name = cleanName(nameTag.text());
    if (!name) return;

    const href = nameTag.attr('href') || '';
    const hashMatch = href.match(/\/([a-fA-F0-9]{40})\//);
    if (!hashMatch) return;
    const magnet = makeMagnet(hashMatch[1], name);

    const sizeText = cols.length > 2 ? $(cols[2]).text().trim() : '';
    const seeders = cols.length > 3 ? parseSeeders($(cols[3]).text()) : 0;

    results.push({
      name,
      size: sizeText || 'N/A',
      size_gb: parseSize(sizeText),
      seeders,
      magnet,
      provider: NAME,
    });
  });
  return results;
}

module.exports = { name: NAME, base_url: BASE, search };

File: scrapers/torrentdownload.py
"""TorrentDownload scraper - HTML scraping."""

import re
from urllib.parse import quote
from scrapers.base import BaseScraper
from scrapers import client, utils


class Source(BaseScraper):
    name = 'TorrentDownload'
    base_url = 'https://www.torrentdownload.info'

    def search(self, query):
        url = f'{self.base_url}/search?q={quote(query)}'
        soup = client.get_soup(url)
        if not soup:
            return []
        results = []
        table = soup.select_one('table.table2')
        if not table:
            # Try alternative layout
            for div in soup.select('div.grey_bar3, div'):
                pass
            return results
        for row in table.select('tr')[1:]:  # skip header row
            cols = row.select('td')
            if len(cols) < 4:
                continue
            # Name from first column link
            name_tag = cols[0].select_one('a')
            name = utils.clean_name(name_tag.get_text()) if name_tag else ''
            if not name:
                continue
            # Get the torrent detail page to find magnet link
            detail_href = name_tag.get('href', '')
            # Size from third column
            size_text = cols[2].get_text(strip=True) if len(cols) > 2 else ''
            # Seeders from fourth column
            seeders = utils.parse_seeders(cols[3].get_text()) if len(cols) > 3 else 0
            # Try to extract hash from the detail link
            hash_match = re.search(r'/([a-fA-F0-9]{40})/', detail_href)
            if hash_match:
                info_hash = hash_match.group(1)
                magnet = self._make_magnet(info_hash, name)
            else:
                continue
            results.append({
                'name': name,
                'size': size_text or 'N/A',
                'size_gb': utils.parse_size(size_text),
                'seeders': seeders,
                'magnet': magnet,
                'provider': self.name,
            })
        return results
